key/src/ed25519_public.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::ops::Deref;
+
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+use sodiumoxide::crypto::sign::{gen_keypair, PublicKey, PUBLICKEYBYTES};
+
+#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Public(PublicKey);
+
+impl Public {
+    pub fn random() -> Self {
+        let (public, _) = gen_keypair();
+        public.into()
+    }
+
+    pub fn from_slice(slice: &[u8]) -> Option<Self> {
+        PublicKey::from_slice(slice).map(Self)
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.as_ref() == [0; PUBLICKEYBYTES]
+    }
+}
+
+impl Default for Public {
+    fn default() -> Self {
+        Public(PublicKey::from_slice(&[0; PUBLICKEYBYTES]).expect("Slice from an array with proper length never fails"))
+    }
+}
+
+impl From<u64> for Public {
+    fn from(integer: u64) -> Self {
+        let for_slice: H256 = integer.into();
+        PublicKey::from_slice(&for_slice).unwrap().into()
+    }
+}
+
+impl FromStr for Public {
+    type Err = crate::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let for_slice = H256::from_str(s).map_err(|e| crate::Error::Custom(format!("{:?}", e)))?;
+        Ok(PublicKey::from_slice(&for_slice).expect("H256 has length 32").into())
+    }
+}
+
+impl Deref for Public {
+    type Target = PublicKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl AsRef<[u8]> for Public {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl From<PublicKey> for Public {
+    fn from(k: PublicKey) -> Self {
+        Public(k)
+    }
+}
+
+impl Encodable for Public {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.append_single_value(&self.as_ref());
```

Comment:
Why does it use `append_single_value`, while you used `append` for `Private`?

Reply:
Sorry, I should have used append_single_value for `Private`.
When I used `append` for the slice, I observed that the encoded `Public` is not separated as an independent list element in the bigger encoded structs.

Reply:
Even, I don't have to implement `Encodable` and `Decodable` for `Private`. I'll modify it.